cfengine3.service, will start other three services

ektich opened the pull request:
During a bootstrap process on systems where systemd is installed CFEngine will attempt to "systemctl restart cfengine3".

This patch introduces cfengine3.service that will restart cf-execd, cf-monitord and cf-serverd.

If cfengine3.service is saved in /etc/systemd/system/ directory it will override the one provided by the CFEngine.

ektich committed:
cfengine3.service, will start other three

During a bootstrap process on systems where systemd is installed
CFEngine will attempt to "systemctl restart cfengine3".
This patch introduces cfengine3.service that will restart cf-execd,
cf-monitord and cf-serverd.

brontolinux commented:
Hi @ektich, thanks for taking the time to contribute.

I acknowledge the need for `cfengine3.service` and I like your idea. I'd be more than happy to add such a unit file (in fact, I had something similar in my first attempts that I had to remove later). Unfortunately, your `cfengine3.service` doesn't honour the enable/disable state of the single services as it is and needs to be fixed before it can be integrated. I'll explain.

On my laptop, only `cf-execd` and `cf-monitord` are enabled; I don't need `cf-serverd` so that unit is disabled. I updated the services files on my system using yours and, in addition, I added `cfengine3.service`, which was of course neither started nor enabled at first:

```
* cfengine3.service - CFEngine 3 daemons
   Loaded: loaded (/usr/local/lib/systemd/system/cfengine3.service; enabled)
   Active: inactive (dead)
     Docs: https://docs.cfengine.com/lts
```

Unfortunately, when I started `cfengine3.service`, it forced `cf-serverd.service` to start because of the dependencies:

```
* cf-serverd.service - CFEngine file server daemon
   Loaded: loaded (/usr/local/lib/systemd/system/cf-serverd.service; disabled)
   Active: active (running) since Sun 2016-05-01 16:31:39 CEST; 6s ago
     Docs: man:cf-serverd
           https://docs.cfengine.com/latest/reference-components-cf-serverd.html
 Main PID: 4405 (cf-serverd)
   CGroup: /system.slice/cf-serverd.service
           `-4405 /var/cfengine/bin/cf-serverd -F

May 01 16:31:39 murray cf-serverd[4405]: notice: Server is starting...
May 01 16:31:39 murray [4405]: CFEngine(server)  Server is starting...
```

I hit the same problem before. If you look into the history of this repository you'll see that I had a `cfengine3.target` unit that I removed for the same reason.

Can you please update the unit file so that it requests, but doesn't require, the separate services to be started, so that the desired state is honoured?

Thanks in advance, ciao!
`-- bronto`

ektich commented:
Just to keep answers in the same place as questions: what about creating symlinks in /etc/systemd/system/cfengine3.service/requires directory to the services that should be started by cfengine3 (and removing Requires= line from the cfengine3.services file)? Don't think about enabling/disabling individual unit files, think about enabling/disabling cfengine3, and listing which parts of cfengine3 should be started when/if it is enabled ( same as /etc/init.d/ + /etc/default/cfengine pair of files do). What do you think about this approach?

brontolinux commented:
Hi @ektich and sorry for the delay, it's a bit messy for me these days...

I see your point (needing a cfengine3.service) and I acknowledge there seem to be no easy way to create an "umbrella service" that starts the three subservices on boot only if they are enabled.

Your approach will work, though I think is a bit "unclean" to me -- you'd be mostly skipping the "API" provided by `systemctl` and doing its job by hand. I think there is a way we can get the best of both worlds (your approach and mine) so that more people can be happy:
- revert the changes in "my" units, so that they don't depend on `cfengine3.service`;
- in `cfengine3.service`, split the `Requires` directive in three separate lines (see next point);
- add a note about `cfengine3.service` in README.md: why that unit file is provided, when it is needed, what one should do if he/she doesn't want to trigger services (that is: comment out the relevant `Requires` lines in the unit file before installing it).
- add a further note explaining that the same result can be accomplished by means of symlinks and killing/commenting out `Requires` lines in `cfengine3.service`, if that is preferable for any reason.

Does that sound good enough for you? Note that I'd asked you to update the README file anyway, so you would not escape that ;-)

ektich committed:
better way of handling dependencies between services

enabling/disabling individual services creates proper "required"
symlinks within /etc/systemd/system/cfengine3.service.requires
directory, making systemd properly start or ignore enabled/disabled
services.

ektich commented:
I think I've got this: with the "RequiredBy" added to individual services and "Requires" removed from the cfengine3.service symlinks within cfengine3.service.required directories will be created/deleted during "systemctl enable/disable" commands, which is what we want in a first place. But as a side effect I had to remove the "WantedBy" lines from individual services, so if the cfengine3.service is not enabled nothing gets started (as individual services are only linked from cfengine3.service, and not from the "multiuser-target". I don't think individual services should be started if they are enabled, but cfengine3.service itself is disabled. Maybe there is a way of specifying some other level of dependencies that will disable individual services when cfengine3 is disabled, but I haven't found it yet.

brontolinux commented on README.md:
````diff
@@ -24,16 +29,16 @@ for Debian](https://cfengine.com/product/community/) and that you'll run
 4. reload the configuration of systemd via ```systemctl daemon-reload```
-5. stop the running CFEngine services, e.g. via ```/etc/init.d/cfengine3 stop```
-6. start the new separate services:<br/> ```systemctl start cf-serverd.service```<br/>and same for cf-execd.service and cf-monitord.service
+5. enable individual services that should be started by cfengine service:<br/>```systemctl enable cf-execd.service```<br/> and same for cf-execd.service and cf-monitord.service
````
Small mistake here: cf-execd is mentioned twice, while cf-serverd is not mentioned.

brontolinux commented:
Hi @ektich 

Apologies it is taking so long :-( I finally got around your changes and they look great. Only one small mistake to fix (see inline comments) and then I'd say it's ready to merge.

**update:** ...or I can fix it myself, whoever is faster ;-)

ektich committed:
typo fix in README.md file

ektich commented:
something like this?

brontolinux commented:
This is now rebased and merged. Unfortunately, probably due to the rebase, git(hub) isn't aware that it is merged. I'll check what I can do to fix that.
